### PR TITLE
Fix mx.arange dtype inference overflow regression

### LINKED ISSUE #4306: [BUG] mx.arange silently returns a saturated int32 array when the range does not fit
### Describe the bug

`mx.arange` silently returns a constant array when the range does not fit in `int32`. Every element comes back as the `int32` limit:

```python
>>> mx.arange(2**40, 2**40 + 3)
array([2147483647, 2147483647, 2147483647], dtype=int32)
>>> mx.arange(-(2**40), -(2**40) + 3)
array([-2147483648, -2147483648, -2147483648], dtype=int32)
```

numpy returns `[1099511627776, 1099511627777, 1099511627778]` as `int64`. Passing the dtype explicitly gives the right answer, so the values are representable, it is only the inferred dtype that is not:

```python
>>> mx.arange(2**40, 2**40 + 3, dtype=mx.int64)
array([1099511627776, 1099511627777, 1099511627778], dtype=int64)
```

This is a change in behaviour from #4255, which landed today. Before that, `Scalar` was `std::variant<bool, int, double>` and this call raised `TypeError`. Now it is accepted and produces the array above, so a case that used to fail loudly returns a wrong answer quietly.

### To Reproduce

```python
import mlx.core as mx
print(mx.arange(2**40, 2**40 + 3))
# array([2147483647, 2147483647, 2147483647], dtype=int32)
```

### Expected behavior

Either `int64`, matching numpy and matching what `mx.array(2**40)` already infers for the same value, or a clear error. Not a silently saturated array.

### Desktop

- OS Version: macOS 15.5 (Darwin 25.5.0)
- Version: 0.32.1.dev20260816+c2bcf47ee, CPU only build

### Additional context

`scalar_to_dtype` in `python/src/ops.cpp` returns `mx::int32` for every integer, now that the variant holds an `int64_t`:

```cpp
mx::Dtype scalar_to_dtype(Scalar s) {
  if (std::holds_alternative<int64_t>(s)) {
    return mx::int32;
```

The reason I am opening this rather than sending a patch is that the obvious fix conflicts with a test #4255 added deliberately:

```python
n = mx.iinfo(mx.int32).max
result = mx.arange(n - 1, n + 3)
self.assertEqual(result.dtype, mx.int32)
self.assertEqual(result.tolist(), [n - 1, n, -2147483648, -2147483647])
```

If `scalar_to_dtype` widens on the value, the way the scalar path in `convert.cpp` already does, then `n + 3` widens that call to `int64` and it returns `[n - 1, n, n + 1, n + 2]` instead of wrapping. That is closer to numpy, but it is a behaviour choice you just made explicitly, so it is yours rather than mine.

Two ways I am happy to implement, whichever you prefer:

1. Widen on value in `scalar_to_dtype`, and update that test to the non wrapping result.
2. Keep the wrapping, and raise when the start value itself cannot be represented in the inferred dtype. That leaves the tested boundary case untouched, since its start does fit, and turns only the meaningless cases into an error.

I am a freshman working through this codebase and I used Claude Code while investigating. Everything above is from a run on this machine.

### PR DESCRIPTION
**Pull request by shoutoutuoadi325**
Fixes #4306.

## Summary

#4255 moved the `Scalar` variant in `python/src/ops.cpp` from `std::variant<bool, int, double>` to `std::variant<bool, int64_t, double>`, but `scalar_to_dtype` still returned `int32` for every integer. Out-of-`int32` arguments therefore stopped raising and `mx.arange` now silently computes in `int32`, where the `double` -> `int32` cast saturates or wraps depending on the backend:

```python
>>> mx.arange(2**40, 2**40 + 3)      # before, and on main today
array([2147483647, 2147483647, 2147483647], dtype=int32)
```

This PR widens the inferred dtype to `int64` when an integer argument does not fit in `int32` (option 1 from the issue), matching numpy and the existing scalar inference in `python/src/convert.cpp` (`mx.array(2**40)` is already `int64`):

```python
>>> mx.arange(2**40, 2**40 + 3)      # after
array([1099511627776, 1099511627777, 1099511627778], dtype=int64)
>>> np.arange(2**40, 2**40 + 3)
array([1099511627776, 1099511627777, 1099511627778])
```

I picked widening over the issue's option 2 (keep wrapping, raise when `start` itself does not fit) because option 2 still leaves cases like `mx.arange(0, 2**40)` silently saturating where numpy widens. Explicit `dtype=` is untouched, and ranges within `int32` keep inferring `int32`.

## Changes

- `python/src/ops.cpp`: `scalar_to_dtype` returns `int64` for integers outside the `int32` range; `arange` docstring mentions the widening.
- `python/tests/test_ops.py`:
  - `test_arange_inferred_dtype` gains positive/negative out-of-range ranges, `int32` boundary values, large and negative steps, the stop-only overload, explicit `dtype=` override, and a float mix.
  - The boundary case added in #4255 now expects `int64` for `mx.arange(n - 1, n + 3)`; the wrapping it used to pin is kept under an explicit `dtype=mx.int32`.
  - `test_arange_overload_dispatch` covers `mx.arange(0, 2**40)` still raising `ValueError` (size limit).

## Test

Built from source on Linux (CPU backend) and ran:

```
pytest python/tests/test_ops.py        # 156 passed, 1 skipped
pytest python/tests                    # passes except test_fft.py::TestFFT::test_fft_grads,
                                       # which fails identically on unmodified main in this
                                       # environment (torch interop, unrelated)
```

Note: arguments still travel to the core `arange` as doubles, so integer arguments beyond 2**53 can lose precision — e.g. `mx.arange(2**53 + 1, 2**53 + 3, dtype=mx.int64)` behaves the same way before and after this change. That is a pre-existing property of the double-based core `arange` and left untouched here.

### CODE REVIEW

**Review by zcbenz — APPROVED** [review]

File: python/src/ops.cpp [hunk 1]
```diff
@@ -1513,7 +1517,8 @@ void init_ops(nb::module_& m) {
           start (float or int, optional): Starting value which defaults to ``0``.
           stop (float or int, optional): Stopping value.
           step (float or int, optional): Increment which defaults to ``1``.
-          dtype (Dtype, optional): Specifies the data type of the output. If unspecified will default to ``float32`` if any of ``start``, ``stop``, or ``step`` are ``float``. Otherwise will default to ``int32``.
+          dtype (Dtype, optional): Specifies the data type of the output. If unspecified will default to ``float32`` if any of ``start``, ``stop``, or ``step`` are ``float``. Otherwise will default to ``int32``, or ``int64``
+            when any of them does not fit in ``int32``.
```
Comment (zcbenz) [inline python/src/ops.cpp]: This is not correct docstring.

Comment (shoutoutuoadi325) [reply python/src/ops.cpp]: Fixed

**Comment by ayaangazali** [comment]
I filed #4306, and zcbenz asked me to check whether this is missing anything. I built the branch and went through it. Short answer: it is complete for what the issue reported, and I would land it.

What I verified:

- The reported case is fixed. `mx.arange(2**40, 2**40 + 3)` gives `int64` and `[1099511627776, 1099511627777, 1099511627778]`, matching numpy.
- You updated the #4255 boundary test rather than leaving it to fail, and keeping the old wrapping available under an explicit `dtype=mx.int32` is a nicer resolution than I proposed in the issue. That preserves the behaviour that PR deliberately locked in while making the inferred case correct.
- `linspace` shares `Scalar` but not the inference, it uses `dtype.value_or(mx::float32)`, so it is unaffected. I checked because they sit next to each other.
- Ordinary cases unchanged: `arange(5)` and `arange(2, 8)` stay `int32`, `arange(0, 1, 0.25)` stays `float32`, negative step still works.
- `test_ops.py` and `test_array.py` pass on the branch.

One thing it does not cover, which I do not think should block it, but is worth knowing since the new `int64` makes the result look trustworthy:

`mx::arange` takes `double` for start, stop and step, so past 2**53 the values are wrong regardless of the dtype:

```python
>>> mx.arange(2**53, 2**53 + 3)
array([9007199254740992, 9007199254740992, 9007199254740992, 9007199254740992], dtype=int64)
>>> mx.arange(2**60, 2**60 + 3)
array([], dtype=int64)
```

numpy gives three distinct values in both. The second is empty because `float(2**60 + 3) - float(2**60)` is exactly `0.0`, so the computed count is zero.

This is pre existing and not caused by this PR: an explicit `dtype=mx.int64` produces the same output on your branch and on main, so it comes from the `double` signature rather than the inference. It was invisible before only because the dtype was clamped to `int32` first. A docstring note, or a follow up that keeps the integer path in integers, would close it.

**Comment by shoutoutuoadi325** [comment]
> I filed #4306, and zcbenz asked me to check whether this is missing anything. I built the branch and went through it. Short answer: it is complete for what the issue reported, and I would land it.
> 
> What I verified:
> 
> * The reported case is fixed. `mx.arange(2**40, 2**40 + 3)` gives `int64` and `[1099511627776, 1099511627777, 1099511627778]`, matching numpy.
> * You updated the [Fix ops rejecting integers larger than INT32_MAX #4255](https://github.com/ml-explore/mlx/pull/4255) boundary test rather than leaving it to fail, and keeping the old wrapping available under an explicit `dtype=mx.int32` is a nicer resolution than I proposed in the issue. That preserves the behaviour that PR deliberately locked in while making the inferred case correct.
> * `linspace` shares `Scalar` but not the inference, it uses `dtype.value_or(mx::float32)`, so it is unaffected. I checked because they sit next to each other.
> * Ordinary cases unchanged: `arange(5)` and `arange(2, 8)` stay `int32`, `arange(0, 1, 0.25)` stays `float32`, negative step still works.
> * `test_ops.py` and `test_array.py` pass on the branch.
> 
> One thing it does not cover, which I do not think should block it, but is worth knowing since the new `int64` makes the result look trustworthy:
> 
> `mx::arange` takes `double` for start, stop and step, so past 2**53 the values are wrong regardless of the dtype:
> 
> ```python
> >>> mx.arange(2**53, 2**53 + 3)
> array([9007199254740992, 9007199254740992, 9007199254740992, 9007199254740992], dtype=int64)
> >>> mx.arange(2**60, 2**60 + 3)
> array([], dtype=int64)
> ```
> 
> numpy gives three distinct values in both. The second is empty because `float(2**60 + 3) - float(2**60)` is exactly `0.0`, so the computed count is zero.
> 
> This is pre existing and not caused by this PR: an explicit `dtype=mx.int64` produces the same output on your branch and on main, so it comes from the `double` signature rather than the inference. It was invisible before only because the dtype was clamped to `int32` first. A docstring note, or a follow up that keeps the integer path in integers, would close it.

I have pushed a new commit to fix the issue of the docstring.
